### DISCUSSION TITLE
(V2) "This field is required" when testing or developing

### DISCUSSION THREAD

**Discussion by ghost** [django-recaptcha/django-recaptcha#335] [Q&A]
The problem is that when ```RECAPTCHA_PUBLIC_KEY``` and ```RECAPTCHA_PRIVATE_KEY``` are removed, as well as when ```SILENCED_SYSTEM_CHECKS = ['captcha.recaptcha_test_key_error']``` is set, the ```captcha.fields.ReCaptchaField``` generates an error "This field is required".

# Code
## settings.py:
```
if 'test' in sys.argv:
    SILENCED_SYSTEM_CHECKS = ['captcha.recaptcha_test_key_error']
else:
    RECAPTCHA_PUBLIC_KEY = os.environ['RECAPTCHA_PUBLIC_KEY']
    RECAPTCHA_PRIVATE_KEY = os.environ['RECAPTCHA_PRIVATE_KEY'] 
```

## forms.py:
```
class RegisterForm(BaseRegisterForm):
    recaptcha = ReCaptchaField(label='')
```

## tests.py:
```
def test_register(self):
    response = self.client.post(reverse('register'), data={
        # Some data for registration
    })

    print(response.content)  # You can find our field with a clear error "This field is required"

    self.assertEquals(response.status_code, 302)  # 200
```

**Comment by ghost** [comment]
@shattie-kiene  Did you get a workaround? I can't use it either. The validation always returns false. Related SO thread: https://stackoverflow.com/questions/62641672/django-form-validation-error-captcha-this-field-is-required

**Comment by ghost** [comment]
No, I didn’t find anything. This is clearly a bug in the code. I can advise you to use 'django-simple-captcha', if it’s not important for you to set recaptcha

**Comment by ghost** [comment]
@shattie-kiene I actually don't want to use simple captcha, but it seems this lib isn't maintained anymore, so I will have to go for a different solution. Let's see.

**Comment by krystofbe** [comment]
You can do something like this in your tests
```
    from unittest import mock

    @mock.patch("captcha.fields.ReCaptchaField.validate")
    def my_test(self, mock_function):
```

and mock the validate function so it doesn't raise an exception

**Comment by VenturaFranklin** [comment]
@axil likely has the longterm solutution with the PR #216 https://github.com/praekelt/django-recaptcha/issues/215#issuecomment-632267812

A work around from @qcaron is in: https://github.com/praekelt/django-recaptcha/issues/157#issuecomment-406301677 also works with a small change. 

I am using the ReCaptchaV3 and had to set return to True
```python
if getattr(settings, "DEBUG", False):
    captcha.clean = lambda x: True
```

**Comment by guettli** [comment]
> @axil likely has the longterm solutution with the PR #216 [#215 (comment)](https://github.com/torchbox/django-recaptcha/issues/215#issuecomment-632267812)
> 
> A work around from @qcaron is in: [#157 (comment)](https://github.com/torchbox/django-recaptcha/issues/157#issuecomment-406301677) also works with a small change.
> 
> I am using the ReCaptchaV3 and had to set return to True
> 
> ```python
> if getattr(settings, "DEBUG", False):
>     captcha.clean = lambda x: True
> ```

This works, but does monkey-patching.  A solution which uses mocking would be better, since mocks gets removed/reset, while monkey-patching changes the code for as long as the interpreter is running.

**Comment by thibaudcolas** [comment]
Sharing another take on the above workarounds for future ref: https://github.com/DjangoGirls/djangogirls/blob/0632a98bf198e3f3f427cb8f9f3334492feda167/core/forms.py#L11C1-L17C37

```py
class BetterReCaptchaField(ReCaptchaField):
    """A ReCaptchaField that always works in DEBUG mode"""

    def clean(self, values):
        if settings.DEBUG:
            return values[0]
        return super().clean(values)
```

**Comment by thibaudcolas** [comment]
Re-reading this I’m not entirely sure whether there’s anything to change in the package here. The test/dev scenario is [documented in the README](https://github.com/django-recaptcha/django-recaptcha#local-development-and-functional-testing).

So I’ll convert to Discussions. Please add more information here or open a new issue if there’s a specific change / documentation improvement to make.
